File: packages/health_connector_hk_ios/ios/health_connector_hk_ios/Sources/health_connector_hk_ios/handlers/ActiveCaloriesBurnedHandler.swift
import Foundation
import HealthKit

/// Handler for active calories burned data (interval quantity type)
///
/// **Data Type:** Active Energy Burned
/// **Record Pattern:** Interval (has startTime and endTime)
/// **Aggregation:** Sum only (cumulative calories)
/// **HealthKit Type:** HKQuantityTypeIdentifier.activeEnergyBurned
///
/// **Usage:**
/// - Reading: Retrieves active calories burned over time intervals
/// - Writing: Records active calories burned during activities
/// - Aggregating: Supports sum (total calories burned)
/// - Pagination: Uses endTime for cursor-based pagination
///
/// **Example:**
/// - Burning 150 kcal during a 30-minute run from 2:00 PM to 2:30 PM
/// - startTime: 2:00 PM, endTime: 2:30 PM, energy: 150 kcal
///
/// **Note:** Active calories exclude basal metabolic rate (BMR).
/// For total calories, use activeEnergyBurned + basalEnergyBurned.

struct ActiveCaloriesBurnedHandler: HealthKitQuantityHandler {
    static var supportedType: HealthDataTypeDto {
        .activeCaloriesBurned
    }
    static var category: HealthKitDataCategory {
        .quantitySample
    }

    static func toDTO(_ sample: HKSample) throws -> HealthRecordDto {
        guard let quantitySample = sample as? HKQuantitySample else {
            throw HealthConnectorErrors.invalidArgument(message: "Expected HKQuantitySample, got \(type(of: sample))")
        }
        guard let dto = quantitySample.toActiveCaloriesBurnedRecordDto() else {
            throw HealthConnectorErrors.invalidArgument(message: "Failed to convert HKQuantitySample to ActiveCaloriesBurnedRecordDto")
        }
        return dto
    }

    static func toHealthKit(_ dto: HealthRecordDto) throws -> HKSample {
        guard let caloriesDto = dto as? ActiveCaloriesBurnedRecordDto else {
            throw HealthConnectorErrors.invalidArgument(
                message: "Expected ActiveCaloriesBurnedRecordDto, got \(type(of: dto))"
            )
        }
        return try caloriesDto.toHealthKit()
    }

    static func getSampleType() -> HKSampleType {
        return HKQuantityType.quantityType(forIdentifier: .activeEnergyBurned)!
    }

    static func extractTimestamp(_ dto: HealthRecordDto) throws -> Int64 {
        guard let caloriesDto = dto as? ActiveCaloriesBurnedRecordDto else {
            throw HealthConnectorErrors.invalidArgument(message: "Expected ActiveCaloriesBurnedRecordDto, got \(type(of: dto))")
        }
        return caloriesDto.endTime
    }

    static func toStatisticsOptions(_ metric: AggregationMetricDto) throws -> HKStatisticsOptions {
        switch metric {
        case .sum:
            return .cumulativeSum
        case .avg, .min, .max, .count:
            throw HealthConnectorErrors.invalidArgument(
                message: "Aggregation metric '\(metric)' not supported for active calories (cumulative data)",
                details: "Only 'sum' is supported"
            )
        }
    }
    
    static func extractAggregateValue(
        from statistics: HKStatistics,
        metric: AggregationMetricDto
    ) throws -> MeasurementUnitDto {
        let quantity: HKQuantity? = switch metric {
        case .sum:
            statistics.sumQuantity()
        case .avg, .min, .max, .count:
            throw HealthConnectorErrors.invalidArgument(
                message: "Aggregation metric '\(metric)' not supported for active calories",
                details: "Only 'sum' is supported"
            )
        }
        
        guard let quantity else {
            throw HealthConnectorErrors.invalidArgument(
                message: "No aggregation result for metric '\(metric)'",
                details: "Statistics returned nil for activeEnergyBurned"
            )
        }
        
        return quantity.toEnergyDto()
    }
}

File: packages/health_connector_hc_android/android/src/main/kotlin/com/phamtunglam/health_connector_hc_android/handlers/ActiveCaloriesBurnedHandler.kt
package com.phamtunglam.health_connector_hc_android.handlers

import androidx.health.connect.client.aggregate.AggregateMetric
import androidx.health.connect.client.aggregate.AggregationResult
import androidx.health.connect.client.records.ActiveCaloriesBurnedRecord
import androidx.health.connect.client.records.Record
import androidx.health.connect.client.units.Energy
import com.phamtunglam.health_connector_hc_android.mappers.health_record_mappers.toDto
import com.phamtunglam.health_connector_hc_android.mappers.health_record_mappers.toHealthConnect
import com.phamtunglam.health_connector_hc_android.mappers.toDto
import com.phamtunglam.health_connector_hc_android.pigeon.ActiveCaloriesBurnedRecordDto
import com.phamtunglam.health_connector_hc_android.pigeon.AggregationMetricDto
import com.phamtunglam.health_connector_hc_android.pigeon.CommonAggregateRequestDto
import com.phamtunglam.health_connector_hc_android.pigeon.HealthDataTypeDto
import com.phamtunglam.health_connector_hc_android.pigeon.HealthRecordDto
import com.phamtunglam.health_connector_hc_android.pigeon.MeasurementUnitDto
import kotlin.reflect.KClass

/**
 * Handler for Active Calories Burned health data type.
 *
 * Characteristics:
 * - Category: Interval record (startTime + endTime)
 * - Aggregation: Supports SUM only
 * - Health Connect Type: ActiveCaloriesBurnedRecord
 */
object ActiveCaloriesBurnedHandler : IntervalRecordHandler, AggregationSupportingHandler<CommonAggregateRequestDto> {
    override val supportedType: HealthDataTypeDto = HealthDataTypeDto.ACTIVE_CALORIES_BURNED

    override fun toDto(record: Record): HealthRecordDto {
        require(record is ActiveCaloriesBurnedRecord) {
            "Expected ActiveCaloriesBurnedRecord, got ${record::class.simpleName}"
        }
        return record.toDto()
    }

    override fun toHealthConnect(dto: HealthRecordDto): Record {
        require(dto is ActiveCaloriesBurnedRecordDto) {
            "Expected ActiveCaloriesBurnedRecordDto, got ${dto::class.simpleName}"
        }
        return dto.toHealthConnect()
    }

    override fun getRecordClass(): KClass<out Record> = ActiveCaloriesBurnedRecord::class

    override fun toAggregateMetric(request: CommonAggregateRequestDto): AggregateMetric<*> =
        when (request.aggregationMetric) {
            AggregationMetricDto.SUM -> ActiveCaloriesBurnedRecord.ACTIVE_CALORIES_TOTAL
            AggregationMetricDto.AVG, AggregationMetricDto.MIN, AggregationMetricDto.MAX, AggregationMetricDto.COUNT ->
                throw UnsupportedOperationException(
                    "Aggregation metric ${request.aggregationMetric} for ActiveCaloriesBurned. Supported: SUM"
                )
        }

    override fun extractAggregateValue(
        aggregationResult: AggregationResult,
        aggregateMetric: AggregateMetric<*>
    ): MeasurementUnitDto {
        val energy = aggregationResult[aggregateMetric] as? Energy
            ?: throw IllegalStateException("Aggregation result for $aggregateMetric is null")
        return energy.toDto()
    }
}
